### PR TITLE
OCPBUGS-13669: skip Azure File installation on Azure StackHub

### PR DESCRIPTION
**Pull request by RomanBednar**

### CODE REVIEW

**Comment by RomanBednar** [comment]
Manually verified with CSO running locally, step 7) is the important one - CSO detects StackHub environment correctly and does not re-deploy Azure File:

1) Install Azure StackHub cluster
```
$ oc get infrastructure/cluster -o yaml | yq '.status.platformStatus.azure.cloudName'
AzureStackCloud
```

2) Set kubeconfig and get version
```
export KUBECONFIG=<path-to-kubeconfig>  
export OPERATOR_IMAGE_VERSION=$(oc get clusterversion/version -o json | jq '.status.desired.version' | tr -d "\"")
```

3) Set required images for deployment
```
export AZURE_DISK_DRIVER_OPERATOR_IMAGE=quay.io/openshift/origin-azure-disk-csi-driver-operator:latest
export DRIVER_IMAGE=quay.io/openshift/origin-azure-disk-csi-driver:latest  
export PROVISIONER_IMAGE=quay.io/openshift/origin-csi-external-provisioner:latest  
export ATTACHER_IMAGE=quay.io/openshift/origin-csi-external-attacher:latest 
export RESIZER_IMAGE=quay.io/openshift/origin-csi-external-resizer:latest  
export SNAPSHOTTER_IMAGE=quay.io/openshift/origin-csi-external-snapshotter:latest  
export NODE_DRIVER_REGISTRAR_IMAGE=quay.io/openshift/origin-csi-node-driver-registrar:latest  
export LIVENESS_PROBE_IMAGE=quay.io/openshift/origin-csi-livenessprobe:latest  
export KUBE_RBAC_PROXY_IMAGE=quay.io/openshift/origin-kube-rbac-proxy:latest  
export CLUSTER_CLOUD_CONTROLLER_MANAGER_OPERATOR_IMAGE=quay.io/openshift/origin-cluster-cloud-controller-manager-operator:latest
```

4) Scale down CVO and CSO  
```
oc scale --replicas=0 deploy/cluster-version-operator -n openshift-cluster-version  
oc scale --replicas=0 deploy/cluster-storage-operator -n openshift-cluster-storage-operator
```

5) Build and run patched CSO locally
```
make
./cluster-storage-operator start --kubeconfig $KUBECONFIG --namespace openshift-cluster-storage-operator
```

6) Verify Azure Disk is recreated by CSO
```
$ oc -n openshift-cluster-csi-drivers delete deployment.apps/azure-disk-csi-driver-operator
deployment.apps "azure-disk-csi-driver-operator" deleted

$ oc -n openshift-cluster-csi-drivers get pod | grep azure-disk-csi-driver-operator
azure-disk-csi-driver-operator-5c7475785-d89rf      1/1     Running            0          7s
```

7) Verify Azure File is not recreated by CSO
```
$ oc -n openshift-cluster-csi-drivers delete deployment.apps/azure-file-csi-driver-operator
deployment.apps "azure-file-csi-driver-operator" deleted

$ oc -n openshift-cluster-csi-drivers get pod | grep azure-file-csi-driver-operator

**Comment by duanwei33** [comment]
Thanks @RomanBednar also did pre-merge test with open PR, just record something here, I think there is no additional test needed from QE side.

```
$ oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.14.0-0.ci.test-2023-09-06-093613-ci-ln-rvnvixt-latest   True        False         124m    Cluster version is 4.14.0-0.ci.test-2023-09-06-093613-ci-ln-rvnvixt-latest

$ oc get clustercsidriver
NAME                 AGE
disk.csi.azure.com   144m

$ oc get sc
NAME                    PROVISIONER          RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
managed-csi (default)   disk.csi.azure.com   Delete          WaitForFirstConsumer   true                   147m

$ oc -n openshift-cluster-csi-drivers get pod | grep file
$ 
```

File: pkg/operator/csidriveroperator/csioperatorclient/azure-file.go [hunk 1]
```diff
@@ -13,6 +13,19 @@ const (
...
 
+func IsNotAzueStackCloud(status *configv1.InfrastructureStatus) bool {
+	if status == nil {
+		return false
+	}
+	// Azure File is not supported on Azure StackHub
+	// https://learn.microsoft.com/en-us/azure-stack/user/azure-stack-acs-differences?view=azs-2206#cheat-sheet-storage-differences
+	// TODO: remove this StatusFilter if Azure File gets support in the future
+	if status.PlatformStatus.Azure.CloudName == configv1.AzureStackCloud {
+		return false
```
Comment (jsafrane) [inline pkg/operator/csidriveroperator/csioperatorclient/azure-file.go:24]: This looks OK for new installations, but what about the existing ones? Suddenly CSO will stop managing the operator. After upgrade, no images will be updated in the operator, running both the operator and CSI driver with potentially insecure binaries.

It should return `false` only when Azure File operator is not running yet, e.g. when its `clustercsidriver` does not exist. It's not a great detection mechanism, but it's better than nothing. Unfortunately, it won't be as simple as filing `StatusFilter`.

**Comment by RomanBednar** [comment]
Manually verified with CSO running locally. CSO will stop deploying Azure File operator only if clustercsidriver is absent:

1) Install Azure StackHub cluster
```
$ oc get infrastructure/cluster -o yaml | yq '.status.platformStatus.azure.cloudName'
AzureStackCloud
```

2) Set kubeconfig and get version
```
export KUBECONFIG=<path-to-kubeconfig>  
export OPERATOR_IMAGE_VERSION=$(oc get clusterversion/version -o json | jq '.status.desired.version' | tr -d "\"")
```

3) Set required images for deployment
```
export AZURE_DISK_DRIVER_OPERATOR_IMAGE=quay.io/openshift/origin-azure-disk-csi-driver-operator:latest
export AZURE_FILE_DRIVER_OPERATOR_IMAGE=quay.io/openshift/origin-azure-file-csi-driver-operator:latest
export AZURE_DISK_DRIVER_IMAGE=quay.io/openshift/origin-azure-disk-csi-driver:latest
export AZURE_FILE_DRIVER_IMAGE=quay.io/openshift/origin-azure-file-csi-driver:latest
export PROVISIONER_IMAGE=quay.io/openshift/origin-csi-external-provisioner:latest  
export ATTACHER_IMAGE=quay.io/openshift/origin-csi-external-attacher:latest 
export RESIZER_IMAGE=quay.io/openshift/origin-csi-external-resizer:latest  
export SNAPSHOTTER_IMAGE=quay.io/openshift/origin-csi-external-snapshotter:latest  
export NODE_DRIVER_REGISTRAR_IMAGE=quay.io/openshift/origin-csi-node-driver-registrar:latest  
export LIVENESS_PROBE_IMAGE=quay.io/openshift/origin-csi-livenessprobe:latest  
export KUBE_RBAC_PROXY_IMAGE=quay.io/openshift/origin-kube-rbac-proxy:latest  
export CLUSTER_CLOUD_CONTROLLER_MANAGER_OPERATOR_IMAGE=quay.io/openshift/origin-cluster-cloud-controller-manager-operator:latest
```

4) Scale down CVO and CSO  
```
oc scale --replicas=0 deploy/cluster-version-operator -n openshift-cluster-version  
oc scale --replicas=0 deploy/cluster-storage-operator -n openshift-cluster-storage-operator
```

5) Build and run patched CSO locally
```
make
./cluster-storage-operator start --kubeconfig $KUBECONFIG --namespace openshift-cluster-storage-operator
```

6) OPTIONAL - delete existing CSO lock
```
oc -n openshift-cluster-storage-operator delete lease/cluster-storage-operator-lock
```

7) Verify Azure Disk is recreated by CSO
```
$ oc -n openshift-cluster-csi-drivers delete deployment.apps/azure-disk-csi-driver-operator
deployment.apps "azure-disk-csi-driver-operator" deleted

$ oc -n openshift-cluster-csi-drivers get pod | grep azure-disk-csi-driver-operator
azure-disk-csi-driver-operator-5c7475785-d89rf      1/1     Running            0          7s
```

8) Verify Azure File is recreated by CSO when installed
```
$ oc -n openshift-cluster-csi-drivers delete deployment.apps/azure-file-csi-driver-operator
deployment.apps "azure-file-csi-driver-operator" deleted

$ oc -n openshift-cluster-csi-drivers get pod | grep azure-file-csi-driver-operator
azure-file-csi-driver-operator-5d6bf99c4b-wk6tf     1/1     Running   0          6s
```

9) Stop CSO

10) Remove Azure File driver
```
$ oc -n openshift-cluster-csi-drivers delete clustercsidriver/file.csi.azure.com deployment.apps/azure-file-csi-driver-operator deployment.apps/azure-file-csi-driver-controller daemonset.apps/azure-file-csi-driver-node service/azure-file-csi-driver-controller-metrics
```

11) Start CSO
```
oc -n openshift-cluster-storage-operator delete lease/cluster-storage-operator-lock
./cluster-storage-operator start --kubeconfig $KUBECONFIG --namespace openshift-cluster-storage-operator
```

12) Verify CSO did not redeploy Azure File operator
```
$ oc get clustercsidriver/file.csi.azure.com
Error from server (NotFound): clustercsidrivers.operator.openshift.io "file.csi.azure.com" not found

$ oc -n openshift-cluster-csi-drivers get pod | grep azure-file-csi-driver-operator
```

File: pkg/operator/csidriveroperator/csioperatorclient/azure-file.go (hunk 1 above)
Comment (RomanBednar) [reply pkg/operator/csidriveroperator/csioperatorclient/azure-file.go]: Good point, but isn't the same issue in `StatusFilter` for IBM VPC? Anyway, I posted a new version that checks the `clustercsidriver` and stops managing Azure File only if it's not present. I had more ideas about the implementation but this seemed ok to me. 

What I'm not sure about is the removal of Azure File in existing StackHub clusters. I just turned off CSO before removing related Azure File objects for testing (see verification below). Do we have some better recommended and well known way for the removal? If not I suppose a release note might be useful. Looking at the bug report it seems pretty much broken so there should not be anything provisioned in existing clusters by Azure File.

Comment (jsafrane) [reply pkg/operator/csidriveroperator/csioperatorclient/azure-file.go]: > but isn't the same issue in StatusFilter for IBM VPC? 

On IBM VPC we check for IBM management cluster and that does not change during upgrade. It should not happen that `isInstalled` gets `true` there.

File: pkg/operatorclient/operatorclient.go [hunk 2]
```diff
@@ -37,6 +38,19 @@ func (c OperatorClient) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv
 	return &instance.Spec.OperatorSpec, &instance.Status.OperatorStatus, instance.ResourceVersion, nil
 }
 
+func (c OperatorClient) IsOperatorInstalled(name string) (bool, error) {
```
Comment (jsafrane) [inline pkg/operatorclient/operatorclient.go:41]: `Operator` in `OperatorClient` refers to CSO, not CSI driver operator. `IsOperatorInstalled` does not really belong here. Use `driver_starter.go` or util.go in the same dir.

File: pkg/operatorclient/operatorclient.go [hunk 3]
```diff
@@ -37,6 +38,19 @@ func (c OperatorClient) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv
...
+	_, err := c.Informers.Operator().V1().ClusterCSIDrivers().Lister().Get(name)
+	if errors.IsNotFound(err) {
+		err = nil
+		return false, err
+	}
+	if err != nil {
+		return false, err
+	}
```
Comment (jsafrane) [inline pkg/operatorclient/operatorclient.go:49]:
```go
if err != nil {
    if errors.IsNotFound(err) {
        return false, nil
    }
    return false, err
}

File: pkg/operator/csidriveroperator/driver_starter.go [hunk 4]
```diff
@@ -196,8 +196,13 @@ func (dsrc *driverStarterCommon) sync(ctx context.Context, syncCtx factory.SyncC
 			return err
 		}
 
+		isInstalled, err := dsrc.commonClients.OperatorClient.IsOperatorInstalled(ctrl.operatorConfig.CSIDriverName)
+		if err != nil {
+			return err
+		}
+
```
Comment (jsafrane) [inline pkg/operator/csidriveroperator/driver_starter.go:203]: Please move it to `!ctrl.running` branch below, so it's called only when needed.

File: pkg/operator/csidriveroperator/csioperatorclient/azure-file.go (hunk 1 above)
Comment (RomanBednar) [reply pkg/operator/csidriveroperator/csioperatorclient/azure-file.go]: Ack, warning added.

File: pkg/operatorclient/operatorclient.go (hunk 2 above)
Comment (RomanBednar) [reply pkg/operatorclient/operatorclient.go]: Done.

File: pkg/operatorclient/operatorclient.go (hunk 3 above)
Comment (RomanBednar) [reply pkg/operatorclient/operatorclient.go]: Done.

File: pkg/operator/csidriveroperator/driver_starter.go (hunk 4 above)
Comment (RomanBednar) [reply pkg/operator/csidriveroperator/driver_starter.go]: Done.

**Pull request merged**